ppx/src/reactjs_jsx_ppx.ml

```diff
@@ -522,33 +535,42 @@ let jsxMapper =
         (fun (arg_label, _) -> "key" = getLabel arg_label)
         nonChildrenProps
     in
-    let jsxExpr, args =
-      let jsxExpr, key, childrenProp =
-        jsxExprAndChildren ~ident:"ReactDOM" ~loc:parentExpLoc ~ctxt mapper
-          ~keyProps children
-      in
-
-      let propsCall =
-        Exp.apply ~loc:parentExpLoc
-          (Exp.ident ~loc:parentExpLoc ~attrs:merlinHideAttrs
-             { loc; txt = Ldot (Lident "ReactDOM", "domProps") })
-          ((match childrenProp with
-           | Some childrenProp ->
-               (labelled "children", childrenProp) :: nonChildrenProps
-           | None -> nonChildrenProps)
-          |> List.map (fun (label, expression) ->
-                 (label, mapper#expression ctxt expression)))
-      in
-      let key_args =
-        match key with
-        | Some (label, key) ->
-            [ (label, mapper#expression ctxt key); (nolabel, unit) ]
-        | None -> []
-      in
-      ( jsxExpr,
-        [ (nolabel, componentNameExpr); (nolabel, propsCall) ] @ key_args )
+    let jsxExpr, key, childrenProp =
+      jsxExprAndChildren ~ident:"ReactDOM" ~loc:parentExpLoc ~ctxt mapper
+        ~keyProps children
+    in
+    let propsCall =
+      Exp.apply ~loc:parentExpLoc
+        (Exp.ident ~loc:parentExpLoc ~attrs:merlinHideAttrs
+           { loc; txt = Ldot (Lident "ReactDOM", "domProps") })
+        ((match childrenProp with
+         | Some childrenProp ->
+             (labelled "children", childrenProp) :: nonChildrenProps
+         | None -> nonChildrenProps)
+        |> List.map (fun (label, expression) ->
+               (label, mapper#expression ctxt expression)))
     in
-    Exp.apply ~loc:parentExpLoc ~attrs jsxExpr args
+    let component = (nolabel, componentNameExpr)
+    and props = (nolabel, propsCall) in
+    match key with
+    | Some (label, key) ->
+        Exp.let_ ~loc Nonrecursive
+          [
+            {
+              pvb_pat = Pat.var { txt = key_var_txt; loc };
+              pvb_expr = mapper#expression ctxt key;
+              pvb_attributes = [];
+              pvb_loc = loc;
+            };
+          ]
+          (Exp.apply ~loc:parentExpLoc ~attrs jsxExpr
```

Comment:
This kind of breaks the rules for "well-formed" locations, because the `apply` will have a larger range (`parentExpLoc`) than the parent `let` expression (which has `loc`). I am afraid that our for now limited tests are not catching a potential regression here.

I think we should at least rename `loc` to `callerLoc` or something else, to prevent this kind of issues in the future.

Reply:
The rename make sense, but I don't understand the issue here. I belive the location is pointed to the right place no?

```
(li ~key:e.path [@JSX]) (* <-- parentExpLoc *)

let reason_react_ppx_key_arg___x = e.path in
ReactDOM.jsxKeyed ~key:reason_react_ppx_key_arg___x "li" (* <-- parentExpLoc *)
```

Reply:
After pressing "Send", I realised that the `let reason_react_ppx_key_arg___x` use the location of the identifier, and this is the reason why when you hover a lowercase component it says "string"

Reply:
In my experience, it doesn't matter that much which is the right location for each node. As long as they are well formed (parents have larger scope than children), things kind of work. One might need to hide some nodes to prioritize others, but the fundamental first step is that the ranges are well formed. When that fails, merlin will fail in most cases.